add API for www.do.de/www.resellerinterface.de

seidler2547 opened the pull request:
First attempt to add the DNS API for Domain-Offensive (www.do.de). I don't know if the API is also available to standard customers, but it is for resellers (hence the domain resellerinterface.de). It seems like the same SOAP API is also used by various other DNS providers under the name Domainrobot.

Parsing XML in shell isn't funny, but I didn't want to introduce a dependency like xmlstarlet.

Neilpang commented:
please fix the CI code format first.

Neilpang commented:
Can you give me a temp user so that I can test the script in all the platforms ?

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+    | grep -F ">${fulldomain}</value>" \
+    | sed -e 's/<item>/\n\0/g' \
+    | grep -F '>id</key>' \
+    | sed -re 's/.*<value[^>]*>([^<]+)<\/value>.*/\1/')"
```
`sed -re` is not supported by solaris

seidler2547 replied:
What's the alternative? Can sed not do regexp at all on Solaris?

Neilpang replied:
do not use sed `-r`,  I think normal regular expression is enough.

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+
+  _all_domains="$(_mktemp)"
+  _dns_do_soap getDomainList
+  echo "${response}" | tr -d "\n\r\t " | grep -Eo 'domain</key><value[^>]+>[^<]+' | sed -re 's/^domain<\/key><value[^>]+>//g' > "${_all_domains}"
```
do not use `grep -Eo`,  please use `_egrep_o()`  function instead.

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+  return 1
+}
+
+_info() {
```
remove `_info()` function

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+  fi
+}
+
+_err() {
```
remove `_err()` function

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+  return 1
+}
+
+_debug() {
```
remove `_debug()` function

seidler2547 commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,167 @@
+  _rr_list="$(echo "${response}" \
+    | tr -d "\n\r\t" \
+    | sed -e 's/<item xsi:type="ns2:Map">/\n/g' \
+    | grep -F ">${fulldomain}</value>" \
```
So could there be a function using the POSIX-compliant version on Solaris? Or is there a function for grepping plain strings?

Neilpang replied:
do not use `grep -F`

seidler2547 commented:
Ok, I have removed `sed -r`. I replaced `grep -F` with `fgrep`, because I think, if you have `egrep`, you should have `fgrep`.

I could give you a test user they gave me some time ago, but obviously it doesn't have any domains associated with it. I can try and ask the company if they are willing to give you a test account.

Neilpang commented:
Please fix the code format first: https://travis-ci.org/Neilpang/acme.sh/jobs/180927758

Is it a must to use fgrep ?  Can you use grep instead ?

seidler2547 commented:
Ok, I didn't see the formatting errors. I have removed fgrep. When using an unknown input as a pattern, I'd rather make sure that it's not some bogus that gets interpreted as regexp. At the very least, the dots in the domain name should escaped, but what do I know what people put in their domain names ;-)

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,144 @@
+        _info "Deleting resource record $_rrid for $_domain"
+        _dns_do_soap deleteRR origin "${_domain}" rrid "${_rrid}"
+        if ! _contains "${response}" '>success<'; then
+          _err "Could not delete resource record for ${_domain}, id ${_rrid}"
```
`return 1` if error happens.

seidler2547 replied:
I don't return here because it would break the loop. I could store the error and `return 1` after the loop? I mean, if the cleanup fails, what happens? It it just communicated to the user that the DNS entries must be removed manually? This loop is here in case previous removal operations failed, to clean up all records that might still exist.

Neilpang replied:
Yes, you can cache the error and return after the loop.

But don't worry if the cleanup fails, just leave the records there.

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,144 @@
+#fulldomain
+dns_do_rm() {
+  fulldomain=$1
+  _cookiejar="$(_mktemp)"
```
Does `_cookiejar` have to be a temp file?
Can we hold the cookie just in memory ?

It's better if we don't use temp file.

seidler2547 replied:
Yes, will change. I was using trial and error because I didn't really know how things are supposed to be.

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,144 @@
+dns_do_add() {
+  fulldomain=$1
+  txtvalue=$2
+  _cookiejar="$(_mktemp)"
```
what about this one ?

Neilpang commented:
please search for `_mktemp` in your file.

Neilpang commented on dnsapi/dns_do.sh:
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env sh
+
+# DNS API for Domain-Offensive / Resellerinterface / Domainrobot
+
```
please add your github repository link here,  so that we can report issues to you once there is any bug found.

seidler2547 commented:
After consulting with their support, their developers say there is currently no test environment for the SOAP API.

Manawyrm commented:
@Neilpang Do you need the account permanently or only for testing purposes?

Neilpang commented:
@Manawyrm 

Just a temp testing account is ok.

Manawyrm commented:
@Neilpang 
Can you send me an e-mail address, so I can securly send you the login credentials to a test useraccount?

Neilpang commented:
@seidler2547  Please fix the shellcheck warnings https://travis-ci.org/Neilpang/acme.sh/jobs/193310358

@Manawyrm  Thank you.  you can send to  `info`**@**acme.sh

Manawyrm commented:
Thank you very much, I have sent you the credentials to test this plugin.

Neilpang commented:
@Manawyrm 
Thank you.  I received your email.  but when I reply you, my email was bounced back.  it seems that your inbox is not valid.

Neilpang commented:
@seidler2547 
Can you please update the steps on how to get started to the readme.md files ?

Manawyrm commented:
@Neilpang That doesn't really sound good. Try replying to github.user@tbspace.de

seidler2547 commented:
@Neilpang about shellcheck: I changed a single character inside a quoted string, why does it complain about setting the headers suddenly?

Neilpang commented:
@seidler2547

Please fix the errors, and update the readme.md files.

Neilpang commented:
@seidler2547 
I was about to merging.  but please rebase and fix the conflicts.

seidler2547 commented:
@Neilpang should be good to go now.
